README.md
# AmpMod
This is a repository for an audio plugin that performs amplitude modulation with triangle and square waves. The modulation frequency ranges from 1 Hz to 25 Hz. This repository is dependent on [IPlug2](https://github.com/iPlug2/iPlug2) framework and its dependencies.

## Triangle Wave
![Triangle Wave](/resources/plugin_image_tri.png)

The image above shows the plugin performing amplitude modulation with triangle wave. The right most knob shows that it is 0% square wave and 100% triangle wave.

## Square Wave
![Square Wave](/resources/plugin_image_sq.png)

The image above shows the plugin performing amplitude modulation with square wave. The right most knob shows that it is 100% square wave and 0% triangle wave.


## Updates
Updated README, triangle wave calculation, and precision on 11/12/2020.


installer/changelog.txt
AmpMod changelog
www.thedeveloperswebsite.com

00/00/00 - v1.00 initial release

AmpMod.h
#pragma once

#include "IPlug_include_in_plug_hdr.h"

const int kNumPresets = 1;

enum EParams
{
    kDepth,
    kFreq,
    kProb,
    kNumParams
};

using namespace iplug;
using namespace igraphics;

class AmpMod final : public Plugin
{
public:
  AmpMod(const InstanceInfo& info);

#if IPLUG_DSP // http://bit.ly/2S64BDd
public:
    void ProcessBlock(sample** inputs, sample** outputs, int nFrames) override;
    void OnReset() override;
    void OnParamChange(int paramIdx) override;
    void initPos();
    void getAmp();
    
private:
    double mDepth = 0.;
    double mFreq = 1.;
    double mProb = 0.;
    
    double mDelaySam1 = 0.;
    double mDelaySam2 = 0.;
    
    int mReadIndex1 = 0;
    int mReadIndex2 = 0;
    
    int mBufferSize1 = 0;
    int mBufferSize2 = 0;
    
    double mAmp1 = 0.;
    double mAmp2 = 0.;
#endif
};


config.h
#define PLUG_NAME "AmpMod"
#define PLUG_MFR "Naoki"
#define PLUG_VERSION_HEX 0x00010000
#define PLUG_VERSION_STR "1.0.0"
#define PLUG_UNIQUE_ID 'v2aS'
#define PLUG_MFR_ID 'Acme'
#define PLUG_URL_STR "https://iplug2.github.io"
#define PLUG_EMAIL_STR "spam@me.com"
#define PLUG_COPYRIGHT_STR "Copyright 2020 Acme Inc"
#define PLUG_CLASS_NAME AmpMod

#define BUNDLE_NAME "AmpMod"
#define BUNDLE_MFR "Naoki"
#define BUNDLE_DOMAIN "com"

#define SHARED_RESOURCES_SUBPATH "AmpMod"

#define PLUG_CHANNEL_IO "1-1 2-2"

#define PLUG_LATENCY 0
#define PLUG_TYPE 0
#define PLUG_DOES_MIDI_IN 0
#define PLUG_DOES_MIDI_OUT 0
#define PLUG_DOES_MPE 0
#define PLUG_DOES_STATE_CHUNKS 0
#define PLUG_HAS_UI 1
#define PLUG_WIDTH 300
#define PLUG_HEIGHT 200
#define PLUG_FPS 60
#define PLUG_SHARED_RESOURCES 0
#define PLUG_HOST_RESIZE 0

#define AUV2_ENTRY AmpMod_Entry
#define AUV2_ENTRY_STR "AmpMod_Entry"
#define AUV2_FACTORY AmpMod_Factory
#define AUV2_VIEW_CLASS AmpMod_View
#define AUV2_VIEW_CLASS_STR "AmpMod_View"

#define AAX_TYPE_IDS 'IEF1', 'IEF2'
#define AAX_TYPE_IDS_AUDIOSUITE 'IEA1', 'IEA2'
#define AAX_PLUG_MFR_STR "Acme"
#define AAX_PLUG_NAME_STR "AmpMod\nIPEF"
#define AAX_PLUG_CATEGORY_STR "Effect"
#define AAX_DOES_AUDIOSUITE 1

#define VST3_SUBCATEGORY "Fx"

#define APP_NUM_CHANNELS 2
#define APP_N_VECTOR_WAIT 0
#define APP_MULT 1
#define APP_COPY_AUV3 0
#define APP_SIGNAL_VECTOR_SIZE 64

#define ROBOTO_FN "Roboto-Regular.ttf"


scripts/make_zip.py
import zipfile, os, fileinput, string, sys

scriptpath = os.path.dirname(os.path.realpath(__file__))
projectpath = os.path.abspath(os.path.join(scriptpath, os.pardir))

MajorStr = ""
MinorStr = "" 
BugfixStr = ""
BUNDLE_NAME = ""

def  main():
  if len(sys.argv) != 2:
    print("Usage: make_zip.py demo(0 or 1)")
    sys.exit(1)
  else:
    demo=int(sys.argv[1])
   
  installer = "\installer\AmpMod Installer.exe"
   
  if demo:
    installer = "\installer\AmpMod Demo Installer.exe"
   
  FILES_TO_ZIP = [
    projectpath + installer,
    projectpath + "\installer\changelog.txt",
    projectpath + "\installer\known-issues.txt",
    projectpath + "\manual\AmpMod manual.pdf" 
  ]

  # extract values from config.h
  for line in fileinput.input(projectpath + "\config.h",inplace=0):
    if "#define PLUG_VERSION_HEX " in line:
      PLUG_VER_STR = string.lstrip(line, "#define PLUG_VERSION_HEX ")
      PLUG_VER = int(PLUG_VER_STR, 16)
      MAJOR = PLUG_VER & 0xFFFF0000
      MAJORSTR = str(MAJOR >> 16)
      MINOR = PLUG_VER & 0x0000FF00
      MINORSTR = str(MINOR >> 8)
      BUGFIXSTR = str(PLUG_VER & 0x000000FF)
    
    if "#define BUNDLE_NAME " in line:
      BUNDLE_NAME = string.lstrip(line, "#define BUNDLE_NAME ")

  FULLVERSIONSTR = MAJORSTR + "." + MINORSTR + "." + BUGFIXSTR

  ZIPNAME = "AmpMod-v" + FULLVERSIONSTR + "-win.zip"
  
  if demo:
    ZIPNAME = "AmpMod-v" + FULLVERSIONSTR + "-win-demo.zip"
  
  zf = zipfile.ZipFile(projectpath + "\installer\/" + ZIPNAME, mode="w")

  for f in FILES_TO_ZIP:
    print("adding " + f)
    zf.write(f, os.path.basename(f));

  zf.close()
  print("wrote ")

if __name__ == '__main__':
  main()

scripts/makedist-mac.sh
#! /bin/sh

BASEDIR=$(dirname $0)

cd $BASEDIR/..

if [ -d build-mac ] 
then
  sudo rm -f -R -f build-mac
fi

#---------------------------------------------------------------------------------------------------------

#variables

DEMO=0
if [ "$1" == "demo" ]
then
DEMO=1
fi

VERSION=`echo | grep PLUG_VERSION_HEX config.h`
VERSION=${VERSION//\#define PLUG_VERSION_HEX }
VERSION=${VERSION//\'}
MAJOR_VERSION=$(($VERSION & 0xFFFF0000))
MAJOR_VERSION=$(($MAJOR_VERSION >> 16)) 
MINOR_VERSION=$(($VERSION & 0x0000FF00))
MINOR_VERSION=$(($MINOR_VERSION >> 8)) 
BUG_FIX=$(($VERSION & 0x000000FF))

FULL_VERSION=$MAJOR_VERSION"."$MINOR_VERSION"."$BUG_FIX

PLUGIN_NAME=`echo | grep BUNDLE_NAME config.h`
PLUGIN_NAME=${PLUGIN_NAME//\#define BUNDLE_NAME }
PLUGIN_NAME=${PLUGIN_NAME//\"}

DMG_NAME=$PLUGIN_NAME-v$FULL_VERSION-mac

if [ $DEMO == 1 ]
then
  DMG_NAME=$DMG_NAME-demo
fi


# work out the paths to the binaries

VST2=`echo | grep VST2_PATH ../../common-mac.xcconfig`
VST2=${VST2//\VST2_PATH = }/$PLUGIN_NAME.vst

VST3=`echo | grep VST3_PATH ../../common-mac.xcconfig`
VST3=${VST3//\VST3_PATH = }/$PLUGIN_NAME.vst3

AU=`echo | grep AU_PATH ../../common-mac.xcconfig`
AU=${AU//\AU_PATH = }/$PLUGIN_NAME.component

APP=`echo | grep APP_PATH ../../common-mac.xcconfig`
APP=${APP//\APP_PATH = }/$PLUGIN_NAME.app

# Dev build folder
AAX=`echo | grep AAX_PATH ../../common-mac.xcconfig`
AAX=${AAX//\AAX_PATH = }/$PLUGIN_NAME.aaxplugin
AAX_FINAL="/Library/Application Support/Avid/Audio/Plug-Ins/$PLUGIN_NAME.aaxplugin"

PKG="installer/build-mac/$PLUGIN_NAME Installer.pkg"
PKG_US="installer/build-mac/$PLUGIN_NAME Installer.unsigned.pkg"

CERT_ID=`echo | grep CERTIFICATE_ID ../../common-mac.xcconfig`
CERT_ID=${CERT_ID//\CERTIFICATE_ID = }

if [ $DEMO == 1 ]
then
  echo "making $PLUGIN_NAME version $FULL_VERSION DEMO mac distribution..."
#   cp "resources/img/AboutBox_Demo.png" "resources/img/AboutBox.png"
else
  echo "making $PLUGIN_NAME version $FULL_VERSION mac distribution..."
#   cp "resources/img/AboutBox_Registered.png" "resources/img/AboutBox.png"
fi

echo ""

#---------------------------------------------------------------------------------------------------------

./scripts/update_installer_version.py $DEMO

echo "touching source to force recompile" 
touch *.cpp

#---------------------------------------------------------------------------------------------------------

#remove existing dist folder
#if [ -d installer/dist ] 
#then
#  rm -R installer/dist
#fi

#mkdir installer/dist

#remove existing binaries
if [ -d $APP ] 
then
  sudo rm -f -R -f $APP
fi

if [ -d $AU ] 
then
  sudo rm -f -R $AU
fi

if [ -d $VST2 ] 
then
  sudo rm -f -R $VST2
fi

if [ -d $VST3 ] 
then
  sudo rm -f -R $VST3
fi

if [ -d "${AAX}" ] 
then
  sudo rm -f -R "${AAX}"
fi

if [ -d "${AAX_FINAL}" ] 
then
  sudo rm -f -R "${AAX_FINAL}"
fi

#---------------------------------------------------------------------------------------------------------

# build xcode project. Change target to build individual formats 

xcodebuild -project ./projects/$PLUGIN_NAME-macOS.xcodeproj -xcconfig ./config/$PLUGIN_NAME-mac.xcconfig DEMO_VERSION=$DEMO -target "All" -configuration Release 2> ./build-mac.log

if [ -s build-mac.log ]
then
  echo "build failed due to following errors:"
  echo ""
  cat build-mac.log
  exit 1
else
 rm build-mac.log
fi

#---------------------------------------------------------------------------------------------------------

#icon stuff - http://maxao.free.fr/telechargements/setfileicon.gz
echo "setting icons"
echo ""
setfileicon resources/$PLUGIN_NAME.icns $AU
setfileicon resources/$PLUGIN_NAME.icns $VST2
setfileicon resources/$PLUGIN_NAME.icns $VST3
setfileicon resources/$PLUGIN_NAME.icns "${AAX}"

#---------------------------------------------------------------------------------------------------------

#strip debug symbols from binaries

echo "stripping binaries"
strip -x $AU/Contents/Resources/plugin.vst3/Contents/MacOS/$PLUGIN_NAME
strip -x $VST2/Contents/MacOS/$PLUGIN_NAME
strip -x $VST3/Contents/MacOS/$PLUGIN_NAME
strip -x $APP/Contents/MacOS/$PLUGIN_NAME
strip -x "${AAX}/Contents/MacOS/$PLUGIN_NAME"

#---------------------------------------------------------------------------------------------------------

#ProTools stuff
echo "copying AAX ${PLUGIN_NAME} from 3PDev to main AAX folder"
sudo cp -p -R "${AAX}" "${AAX_FINAL}"
mkdir "${AAX_FINAL}/Contents/Factory Presets/"

echo "code sign AAX binary"
/Applications/PACEAntiPiracy/Eden/Fusion/Current/bin/wraptool sign --verbose --account XXXX --wcguid XXXX --signid "Developer ID Application: ""${CERT_ID}" --in "${AAX_FINAL}" --out "${AAX_FINAL}"

#---------------------------------------------------------------------------------------------------------

#Mac AppStore stuff

#xcodebuild -project $PLUGIN_NAME.xcodeproj -xcconfig $PLUGIN_NAME.xcconfig -target "APP" -configuration Release 2> ./build-mac.log

#echo "code signing app for appstore"
#echo ""
#codesign -f -s "3rd Party Mac Developer Application: ""${CERT_ID}" $APP --entitlements resources/$PLUGIN_NAME.entitlements
 
#echo "building pkg for app store"
#echo ""
#productbuild \
#     --component $APP /Applications \
#     --sign "3rd Party Mac Developer Installer: ""${CERT_ID}" \
#     --product "/Applications/$PLUGIN_NAME.app/Contents/Info.plist" installer/$PLUGIN_NAME.pkg

#---------------------------------------------------------------------------------------------------------

#10.8 Gatekeeper/Developer ID stuff

#echo "code app binary for Gatekeeper on 10.8"
#echo ""
#codesign -f -s "Developer ID Application: ""${CERT_ID}" $APP

#---------------------------------------------------------------------------------------------------------

# installer, uses Packages http://s.sudre.free.fr/Software/Packages/about.html
sudo sudo rm -R -f installer/$PLUGIN_NAME-mac.dmg

echo "building installer"
echo ""
packagesbuild installer/$PLUGIN_NAME.pkgproj

echo "code-sign installer for Gatekeeper on 10.8"
echo ""
mv "${PKG}" "${PKG_US}"
productsign --sign "Developer ID Installer: ""${CERT_ID}" "${PKG_US}" "${PKG}"
                   
rm -R -f "${PKG_US}"

#set installer icon
setfileicon resources/$PLUGIN_NAME.icns "${PKG}"

#---------------------------------------------------------------------------------------------------------

# dmg, can use dmgcanvas http://www.araelium.com/dmgcanvas/ to make a nice dmg

echo "building dmg"
echo ""

if [ -d installer/$PLUGIN_NAME.dmgCanvas ]
then
  dmgcanvas installer/$PLUGIN_NAME.dmgCanvas installer/$DMG_NAME.dmg
else
  hdiutil create installer/$DMG_NAME.dmg -format UDZO -srcfolder installer/build-mac/ -ov -anyowners -volname $PLUGIN_NAME
fi

sudo rm -R -f installer/build-mac/

#---------------------------------------------------------------------------------------------------------
# zip

# echo "copying binaries..."
# echo ""
# cp -R $AU installer/dist/$PLUGIN_NAME.component
# cp -R $VST2 installer/dist/$PLUGIN_NAME.vst
# cp -R $VST3 installer/dist/$PLUGIN_NAME.vst3
# cp -R $AAX installer/dist/$PLUGIN_NAME.aaxplugin
# cp -R $APP installer/dist/$PLUGIN_NAME.app
# 
# echo "zipping binaries..."
# echo ""
# ditto -c -k installer/dist installer/$PLUGIN_NAME-mac.zip
# rm -R installer/dist

#---------------------------------------------------------------------------------------------------------

if [ $DEMO == 1 ]
then
git checkout installer/AmpMod.iss
git checkout installer/AmpMod.pkgproj
git checkout resources/img/AboutBox.png
fi

echo "done"


scripts/makedist-web.sh
#!/bin/bash

# makedist-web.sh builds a Web version of an iPlug2 project using emscripten
# it copies a template folder from the iPlug2 tree and does a find and replace on various JavaScript and HTML files
# arguments:
# 1st argument : either "on", "off" or "ws" - this specifies whether emrun is used to launch a server and browser after compilation. "ws" builds the project in websocket mode, without the WAM stuff
# 2nd argument : site origin -
# 3rd argument : browser - either "chrome", "safari", "firefox" - if you want to launch a browser other than chrome, you must specify the correct origin for argument #2

SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
IPLUG2_ROOT=../../..
PROJECT_ROOT=$SCRIPT_DIR/..
IPLUG2_ROOT=$SCRIPT_DIR/$IPLUG2_ROOT

PROJECT_NAME=AmpMod
WEBSOCKET_MODE=0
EMRUN_BROWSER=chrome
LAUNCH_EMRUN=1
EMRUN_SERVER=1
EMRUN_SERVER_PORT=8001
SITE_ORIGIN="/"

cd $PROJECT_ROOT

if [ "$1" = "ws" ]; then
  LAUNCH_EMRUN=0
  WEBSOCKET_MODE=1
elif [ "$1" = "off" ]; then
  LAUNCH_EMRUN=0
fi

if [ "$#" -eq 2 ]; then
  SITE_ORIGIN=${2}
fi

if [ "$#" -eq 3 ]; then
  EMRUN_BROWSER=${3}
fi

# check to see if the build web folder has its own git repo
if [ -d build-web/.git ]
then
  # if so trash only the scripts folder
  if [ -d build-web/scripts ]; then rm -r build-web/scripts; fi
else
  # otherwise trash the whole build-web folder
  if [ -d build-web ]; then rm -r build-web; fi

  mkdir build-web
fi

mkdir build-web/scripts

echo BUNDLING RESOURCES -----------------------------

cd build-web

if [ -f imgs.js ]; then rm imgs.js; fi
if [ -f imgs@2x.js ]; then rm imgs@2x.js; fi
if [ -f svgs.js ]; then rm svgs.js; fi
if [ -f fonts.js ]; then rm fonts.js; fi

#package fonts
FOUND_FONTS=0
if [ "$(ls -A ../resources/fonts/*.ttf)" ]; then
  FOUND_FONTS=1
  python $EMSDK/upstream/emscripten/tools/file_packager.py fonts.data --preload ../resources/fonts/ --exclude *DS_Store --js-output=fonts.js
fi

#package svgs
FOUND_SVGS=0
if [ "$(ls -A ../resources/img/*.svg)" ]; then
  FOUND_SVGS=1
  python $EMSDK/upstream/emscripten/tools/file_packager.py svgs.data --preload ../resources/img/ --exclude *.png --exclude *DS_Store --js-output=svgs.js
fi

#package @1x pngs
FOUND_PNGS=0
if [ "$(ls -A ../resources/img/*.png)" ]; then
  FOUND_PNGS=1
  python $EMSDK/upstream/emscripten/tools/file_packager.py imgs.data --use-preload-plugins --preload ../resources/img/ --use-preload-cache --indexedDB-name="/$PROJECT_NAME_pkg" --exclude *DS_Store --exclude  *@2x.png --exclude  *.svg >> imgs.js
fi

# package @2x pngs into separate .data file
FOUND_2XPNGS=0
if [ "$(ls -A ../resources/img/*@2x*.png)" ]; then
  FOUND_2XPNGS=1
  mkdir ./2x/
  cp ../resources/img/*@2x* ./2x
  python $EMSDK/upstream/emscripten/tools/file_packager.py imgs@2x.data --use-preload-plugins --preload ./2x@/resources/img/ --use-preload-cache --indexedDB-name="/$PROJECT_NAME_pkg" --exclude *DS_Store >> imgs@2x.js
  rm -r ./2x
fi

cd ..

if [ "$WEBSOCKET_MODE" -eq "0" ]; then
  echo MAKING  - WAM WASM MODULE -----------------------------
  cd $PROJECT_ROOT/projects
  emmake make --makefile $PROJECT_NAME-wam-processor.mk

  if [ $? -ne "0" ]; then
    echo IPlugWAM WASM compilation failed
    exit 1
  fi

  cd $PROJECT_ROOT/build-web/scripts

  # prefix the -wam.js script with scope
  echo "AudioWorkletGlobalScope.WAM = AudioWorkletGlobalScope.WAM || {}; AudioWorkletGlobalScope.WAM.$PROJECT_NAME = { ENVIRONMENT: 'WEB' };" > $PROJECT_NAME-wam.tmp.js;
  cat $PROJECT_NAME-wam.js >> $PROJECT_NAME-wam.tmp.js
  mv $PROJECT_NAME-wam.tmp.js $PROJECT_NAME-wam.js
  
  # copy in WAM SDK and AudioWorklet polyfill scripts
  cp $IPLUG2_ROOT/Dependencies/IPlug/WAM_SDK/wamsdk/*.js .
  cp $IPLUG2_ROOT/Dependencies/IPlug/WAM_AWP/*.js .

  # copy in template scripts
  cp $IPLUG2_ROOT/IPlug/WEB/Template/scripts/IPlugWAM-awn.js $PROJECT_NAME-awn.js
  cp $IPLUG2_ROOT/IPlug/WEB/Template/scripts/IPlugWAM-awp.js $PROJECT_NAME-awp.js

  # replace NAME_PLACEHOLDER in the template -awn.js and -awp.js scripts
  sed -i.bak s/NAME_PLACEHOLDER/$PROJECT_NAME/g $PROJECT_NAME-awn.js
  sed -i.bak s/NAME_PLACEHOLDER/$PROJECT_NAME/g $PROJECT_NAME-awp.js

  # replace ORIGIN_PLACEHOLDER in the template -awn.js script
  sed -i.bak s,ORIGIN_PLACEHOLDER,$SITE_ORIGIN,g $PROJECT_NAME-awn.js

  rm *.bak
else
  echo "WAM not being built in websocket mode"
fi

cd $PROJECT_ROOT/build-web

# copy in the template HTML - comment this out if you have customised the HTML
cp $IPLUG2_ROOT/IPlug/WEB/Template/index.html index.html
sed -i.bak s/NAME_PLACEHOLDER/$PROJECT_NAME/g index.html

if [ $FOUND_FONTS -eq "0" ]; then sed -i.bak s/'<script async src="fonts.js"><\/script>'/'<!--<script async src="fonts.js"><\/script>-->'/g index.html; fi
if [ $FOUND_SVGS -eq "0" ]; then sed -i.bak s/'<script async src="svgs.js"><\/script>'/'<!--<script async src="svgs.js"><\/script>-->'/g index.html; fi
if [ $FOUND_PNGS -eq "0" ]; then sed -i.bak s/'<script async src="imgs.js"><\/script>'/'<!--<script async src="imgs.js"><\/script>-->'/g index.html; fi
if [ $FOUND_2XPNGS -eq "0" ]; then sed -i.bak s/'<script async src="imgs@2x.js"><\/script>'/'<!--<script async src="imgs@2x.js"><\/script>-->'/g index.html; fi
if [ $WEBSOCKET_MODE -eq "1" ]; then
  cp $IPLUG2_ROOT/Dependencies/IPlug/WAM_SDK/wamsdk/wam-controller.js scripts/wam-controller.js
  cp $IPLUG2_ROOT/IPlug/WEB/Template/scripts/websocket.js scripts/websocket.js
  sed -i.bak s/'<script src="scripts\/audioworklet.js"><\/script>'/'<!--<script src="scripts\/audioworklet.js"><\/script>-->'/g index.html;
  sed -i.bak s/'let WEBSOCKET_MODE=false;'/'let WEBSOCKET_MODE=true;'/g index.html;
else
  sed -i.bak s/'<script src="scripts\/websocket.js"><\/script>'/'<!--<script src="scripts\/websocket.js"><\/script>-->'/g index.html;

  # update the i/o details for the AudioWorkletNodeOptions parameter, based on config.h channel io str
  MAXNINPUTS=$(python $IPLUG2_ROOT/Scripts/parse_iostr.py "$PROJECT_ROOT" inputs)
  MAXNOUTPUTS=$(python $IPLUG2_ROOT/Scripts/parse_iostr.py "$PROJECT_ROOT" outputs)

  if [ $MAXNINPUTS -eq "0" ]; then 
    MAXNINPUTS="";
    sed -i.bak '181,203d' index.html; # hack to remove GetUserMedia() from code, and allow WKWebKitView usage for instruments
  fi
  sed -i.bak s/"MAXNINPUTS_PLACEHOLDER"/"$MAXNINPUTS"/g index.html;
  sed -i.bak s/"MAXNOUTPUTS_PLACEHOLDER"/"$MAXNOUTPUTS"/g index.html;
fi

rm *.bak

# copy the style & WAM favicon
mkdir styles
cp $IPLUG2_ROOT/IPlug/WEB/Template/styles/style.css styles/style.css
cp $IPLUG2_ROOT/IPlug/WEB/Template/favicon.ico favicon.ico

echo MAKING  - WEB WASM MODULE -----------------------------

cd $PROJECT_ROOT/projects

emmake make --makefile $PROJECT_NAME-wam-controller.mk EXTRA_CFLAGS=-DWEBSOCKET_CLIENT=$WEBSOCKET_MODE

if [ $? -ne "0" ]; then
  echo IPlugWEB WASM compilation failed
  exit 1
fi

cd $PROJECT_ROOT/build-web

# print payload
echo payload:
find . -maxdepth 2 -mindepth 1 -name .git -type d \! -prune -o \! -name .DS_Store -type f -exec du -hs {} \;

# launch emrun
if [ "$LAUNCH_EMRUN" -eq "1" ]; then
  if [ "$EMRUN_SERVER" -eq "0" ]; then
    emrun --browser $EMRUN_BROWSER --no_server --port=$EMRUN_SERVER_PORT index.html
  else
    emrun --browser $EMRUN_BROWSER --no_emrun_detect index.html
  fi
else
  echo "Not running emrun"
fi


scripts/update_version.py
#!/usr/bin/python

# this script will update the versions in plist and installer files to match that in config.h

import plistlib, os, datetime, fileinput, glob, sys, string, shutil

scriptpath = os.path.dirname(os.path.realpath(__file__))
projectpath = os.path.abspath(os.path.join(scriptpath, os.pardir))

IPLUG2_ROOT = "../../.."

sys.path.insert(0, os.path.join(os.getcwd(), IPLUG2_ROOT + '/Scripts'))

from parse_config import parse_config, parse_xcconfig

def replacestrs(filename, s, r):
  files = glob.glob(filename)
  
  for line in fileinput.input(files,inplace=1):
    string.find(line, s)
    line = line.replace(s, r)
    sys.stdout.write(line)

def main():

  config = parse_config(projectpath)
    
  today = datetime.date.today()
  
  CFBundleGetInfoString = config['BUNDLE_NAME'] + " v" + config['FULL_VER_STR'] + " " + config['PLUG_COPYRIGHT_STR']
  CFBundleVersion = config['FULL_VER_STR']

  print "update_version.py - setting version to " + config['FULL_VER_STR']
  print "Updating plist version info..."
  
  plistpath = scriptpath + "/resources/AmpMod-VST2-Info.plist"
  vst2 = plistlib.readPlist(plistpath)
  vst2['CFBundleGetInfoString'] = CFBundleGetInfoString
  vst2['CFBundleVersion'] = CFBundleVersion
  vst2['CFBundleShortVersionString'] = CFBundleVersion
  plistlib.writePlist(vst2, plistpath)
  replacestrs(plistpath, "//Apple//", "//Apple Computer//");
  
  plistpath = scriptpath + "/resources/AmpMod-AU-Info.plist"
  au = plistlib.readPlist(plistpath)
  au['CFBundleGetInfoString'] = CFBundleGetInfoString
  au['CFBundleVersion'] = CFBundleVersion
  au['CFBundleShortVersionString'] = CFBundleVersion
  plistlib.writePlist(au, plistpath)
  replacestrs(plistpath, "//Apple//", "//Apple Computer//");
  
  plistpath = scriptpath + "/resources/AmpMod-VST3-Info.plist"
  vst3 = plistlib.readPlist(plistpath)
  vst3['CFBundleGetInfoString'] = CFBundleGetInfoString
  vst3['CFBundleVersion'] = CFBundleVersion
  vst3['CFBundleShortVersionString'] = CFBundleVersion
  plistlib.writePlist(vst3, plistpath)
  replacestrs(plistpath, "//Apple//", "//Apple Computer//");
  
  plistpath = scriptpath + "/resources/AmpMod-macOS-Info.plist"
  app = plistlib.readPlist(plistpath)
  app['CFBundleGetInfoString'] = CFBundleGetInfoString
  app['CFBundleVersion'] = CFBundleVersion
  app['CFBundleShortVersionString'] = CFBundleVersion
  plistlib.writePlist(app, plistpath)
  replacestrs(plistpath, "//Apple//", "//Apple Computer//");
  
  plistpath = scriptpath + "/resources/AmpMod-AAX-Info.plist"
  aax = plistlib.readPlist(plistpath)
  aax['CFBundleGetInfoString'] = CFBundleGetInfoString
  aax['CFBundleVersion'] = CFBundleVersion
  aax['CFBundleShortVersionString'] = CFBundleVersion
  plistlib.writePlist(aax, plistpath)
  replacestrs(plistpath, "//Apple//", "//Apple Computer//");

  print "Updating Mac Installer version info..."
  
  plistpath = scriptpath + "/installer/AmpMod.pkgproj"
  installer = plistlib.readPlist(plistpath)
  
  for x in range(0,5):
    installer['PACKAGES'][x]['PACKAGE_SETTINGS']['VERSION'] = config['FULL_VER_STR']
  
  plistlib.writePlist(installer, plistpath)
  replacestrs(plistpath, "//Apple//", "//Apple Computer//");
  
  print "Updating Windows Installer version info..."
  
  for line in fileinput.input(scriptpath + "/installer/AmpMod.iss",inplace=1):
    if "AppVersion" in line:
      line="AppVersion=" + config['FULL_VER_STR'] + "\n"
    sys.stdout.write(line)

if __name__ == '__main__':
  main()


AmpMod.cpp
#include "AmpMod.h"
#include "IPlug_include_in_plug_src.h"
#include "IControls.h"
#include <math.h>

AmpMod::AmpMod(const InstanceInfo& info)
: Plugin(info, MakeConfig(kNumParams, kNumPresets))
{
    GetParam(kDepth)->InitDouble("Depth", 50., 0., 100., 0.01, "%");
    GetParam(kFreq)->InitFrequency("Frequency", 1., 1., 25., 0.1);
    GetParam(kProb)->InitDouble("Tri-Sq", 50., 0., 100., 0.1, "%");

#if IPLUG_EDITOR // http://bit.ly/2S64BDd
    mMakeGraphicsFunc = [&]() {
        return MakeGraphics(*this, PLUG_WIDTH, PLUG_HEIGHT, PLUG_FPS, GetScaleForScreen(PLUG_HEIGHT));
    };
    
    mLayoutFunc = [&](IGraphics* pGraphics) {
        pGraphics->AttachCornerResizer(EUIResizerMode::Scale, false);
        pGraphics->AttachPanelBackground(COLOR_WHITE);
        pGraphics->LoadFont("Roboto-Regular", ROBOTO_FN);
        const IRECT b = pGraphics->GetBounds();
        pGraphics->AttachControl(new ITextControl(b.GetMidVPadded(50).GetVShifted(-60), "Amplitude Modulation", IText(30)));
        
        pGraphics->AttachControl(new IVKnobControl(b.GetCentredInside(100).GetVShifted(20).GetHShifted(0), kDepth));
        pGraphics->AttachControl(new IVKnobControl(b.GetCentredInside(100).GetVShifted(20).GetHShifted(-90), kFreq));
        pGraphics->AttachControl(new IVKnobControl(b.GetCentredInside(100).GetVShifted(20).GetHShifted(90), kProb));
  };
#endif
}

#if IPLUG_DSP
void AmpMod::ProcessBlock(sample** inputs, sample** outputs, int nFrames)
{
    sample* in1 = inputs[0];
    sample* in2 = inputs[1];
    sample* out1 = outputs[0];
    sample* out2 = outputs[1];
    
    mDepth = GetParam(kDepth)->Value();
    double depth = (100. + mDepth) / 100.;
    
    for (int s = 0; s < nFrames; ++s, ++in1, ++in2, ++out1, ++out2)
    {
        getAmp();
        
        // Multiply buffer coefficients with output
        *out1 = (depth * mAmp1) * *in1;
        *out2 = (depth * mAmp2) * *in2;

        //increment the read index, wrapping if it goes out of bounds.
        mReadIndex1++;
        if(mReadIndex1 >= mBufferSize1) {
            mReadIndex1 = 0;
        }
        mReadIndex2++;
        if(mReadIndex2 >= mBufferSize2) {
            mReadIndex2 = 0;
        }
    }
}


void AmpMod::OnReset() {
    TRACE;
    
    mReadIndex1 = 0;
    mReadIndex2 = 0;
    mFreq = 1.;
    getAmp();
}

void AmpMod::initPos() {
    mFreq = GetParam(kFreq)->Value();
    mFreq = floor(mFreq * 10 + 0.5)/10;
    
    // Find the global position in the project
    int pos1 = (int) GetSamplePos();
    int pos2 = (int) GetSamplePos();
    
    // Compute Period
    mBufferSize1 = floor(GetSampleRate() / mFreq);
    mBufferSize2 = floor(GetSampleRate() / mFreq);
    
    // Get position with respect to the modulation wave
    if (mBufferSize1 != 0 && mBufferSize2 != 0) {
        mReadIndex1 = pos1 % mBufferSize1;
        mReadIndex2 = pos2 % mBufferSize2;
    }
    else {
        mReadIndex1 = 0;
        mReadIndex2 = 0;
    }
}

void AmpMod::getAmp() {
    mProb = GetParam(kProb)->Value()/100;
    
    // Get triangle and square wave amplitude
    if (mReadIndex1 < floor(mBufferSize1/2)) {
        mAmp1 = mProb*abs((2/M_PI)*asin(sin(M_PI*mFreq*mReadIndex1/GetSampleRate()))) + (1-mProb);
        mAmp2 = mProb*abs((2/M_PI)*asin(sin(M_PI*mFreq*mReadIndex2/GetSampleRate()))) + (1-mProb);
    }
    else {
      mAmp1 = mProb*abs((2/M_PI)*asin(sin(M_PI*mFreq*mReadIndex1/GetSampleRate())));
      mAmp2 = mProb*abs((2/M_PI)*asin(sin(M_PI*mFreq*mReadIndex2/GetSampleRate())));
    }
}

void AmpMod::OnParamChange(int paramIdx) {
    mFreq = 1.;
    initPos();
    getAmp();
}
#endif
